### ISSUE TITLE
Assert in ebpfcore.sys (in _ebpf_link_instance_invoke_batch_begin())

### ISSUE THREAD

**Issue by dv-msft** [microsoft/ebpf-for-windows#2466]
### Describe the bug

Running a test from the 'multi-threaded kernel mode stress' test suite triggers an assert in ebpfcore.  The behavior is consistent (5/5  attempts) and reproduces within 10 - 15 seconds into the test run.

### Steps taken to reproduce bug

(With kd attached), run the kernel mode multi-threaded stress test with the following command line:
`.\ebpf_stress_tests_km -tt=8 -td=1 -erd=250 native_invoke_program_restart_extension_v4_test`

### Expected behavior

The rest should pass successfully.

### Actual outcome

The test triggers an assert in `ebpfcore.sys` within the first minute of the test run.

Kernel stack trace:

```
0: kd> k
 # Child-SP          RetAddr               Call Site
00 ffffab03`9782d160 fffff805`42256e05     nt!RtlAssert+0xee [minkernel\ntos\rtl\assert.c @ 132] 
01 ffffab03`9782d6c0 fffff805`42256b38     ebpfcore!_ebpf_link_instance_invoke_batch_begin+0x1d5 [D:\wrk\ebpf-for-windows\libs\execution_context\ebpf_link.c @ 407] 
02 ffffab03`9782d7b0 fffff805`51277054     ebpfcore!_ebpf_link_instance_invoke+0x48 [D:\wrk\ebpf-for-windows\libs\execution_context\ebpf_link.c @ 363] 
03 ffffab03`9782d820 fffff805`5127e2f8     netebpfext!net_ebpf_extension_hook_invoke_program+0x74 [D:\wrk\ebpf-for-windows\netebpfext\net_ebpf_ext_hook_provider.c @ 220] 
04 ffffab03`9782d920 fffff805`3ec39c2a     netebpfext!net_ebpf_extension_sock_addr_redirect_connection_classify+0x5c8 [D:\wrk\ebpf-for-windows\netebpfext\net_ebpf_ext_sock_addr.c @ 1624] 
05 ffffab03`9782db60 fffff805`3e2fc737     tcpip+0xc9c2a
06 ffffab03`9782dbb0 fffff805`3e2fa967     NETIO!KfdClassify+0x2337
07 ffffab03`9782dd30 fffff805`3eb971ff     NETIO!KfdClassify+0x567
08 ffffab03`9782e100 fffff805`3eb96974     tcpip+0x271ff
09 ffffab03`9782ea20 fffff805`3eb963bf     tcpip+0x26974
0a ffffab03`9782eb60 fffff805`3eb95d19     tcpip+0x263bf
0b ffffab03`9782ecf0 fffff805`3948f18a     tcpip+0x25d19
```

Assert condition: `ebpf_assert(link->state == EBPF_LINK_STATE_ATTACHED);`

```
11: kd> dx link
link                 : 0xffffd0054f246680 [Type: _ebpf_link *]
    [+0x000] object           [Type: _ebpf_core_object]
    [+0x058] program          : 0xffffd0054f571b20 [Type: _ebpf_program *]
    [+0x060] attach_type      : {A82E37B1-AEE7-11EC-9A30-18602489BEEE} [Type: _GUID]
    [+0x070] bpf_attach_type  : BPF_CGROUP_INET4_CONNECT (3) [Type: bpf_attach_type]
    [+0x074] link_type        : BPF_LINK_TYPE_CGROUP (2) [Type: bpf_link_type]
    [+0x078] program_type     : {92EC8E39-AEEC-11EC-9A30-18602489BEEE} [Type: _GUID]
    [+0x088] client_data      [Type: _ebpf_extension_data]
    [+0x0a0] extension_client_context : 0xffffd005440fb540 [Type: _ebpf_extension_client *]
    [+0x0a8] attach_lock      : 0x0 [Type: unsigned __int64]
    [+0x0b0] state            : EBPF_LINK_STATE_ATTACHING (1) [Type: _ebpf_link_state]
    [+0x0b8] provider_binding_context : 0xffffd0054eed3710 [Type: void *]
```

**Comment by Alan-Jowett** [comment]
Overactive assert. It asserts on the link state outside of the lock.

**Comment by dv-msft** [comment]
Fixed in PR #2434
